src/bin/cli/logs/sentry/download.rs

```diff
@@ -0,0 +1,179 @@
+use super::Error;
+use catalyst_toolbox::logs::sentry::{LazySentryLogs, RawLog, SentryLogClient};
+use jcli_lib::utils::io::open_file_write;
+
+use std::path::PathBuf;
+
+use chrono::{DateTime, FixedOffset};
+use std::str::FromStr;
+use structopt::StructOpt;
+use url::Url;
+
+const DATE_TIME_TAG: &str = "dateCreated";
+
+pub enum Mode {
+    Full,
+    Latest,
+}
+
+#[derive(StructOpt)]
+pub struct DateFilter {
+    // Optional [From, To] date range, start
+    #[structopt(long, parse(try_from_str = DateTime::parse_from_rfc3339))]
+    from: Option<DateTime<FixedOffset>>,
+    // Optional [From, To] date range, end
+    #[structopt(long, parse( try_from_str = DateTime::parse_from_rfc3339))]
+    to: Option<DateTime<FixedOffset>>,
+}
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub struct Download {
+    /// Sentry logs url
+    #[structopt(long)]
+    url: Url,
+
+    /// Sentry access token
+    #[structopt(long)]
+    token: String,
+
+    /// Path to the ouput file, will be overwritten if exists.
+    #[structopt(long)]
+    out: PathBuf,
+
+    /// Choose between latest logs or full logs download
+    #[structopt(long, default_value = "latest")]
+    mode: Mode,
+
+    #[structopt(flatten)]
+    dates: DateFilter,
+
+    /// Size of retrieved logs, only used in "full" Mode
+    #[structopt(long, default_value = "1000")]
+    chunk_size: usize,
+}
+
+impl FromStr for Mode {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "f" | "full" => Ok(Self::Full),
+            "l" | "latest" => Ok(Self::Latest),
+            _ => Err(format!(
+                "Could not parse Mode {}. Any of 'f', 'full', 'l' or 'latest' is required",
+                s
+            )),
+        }
+    }
+}
+
+impl Download {
+    pub fn exec(self) -> Result<(), Error> {
+        let Self {
+            url,
+            token,
+            out,
+            mode,
+            dates,
+            chunk_size,
+        } = self;
+        let dates = flip_dates_if_wrong(dates);
+
+        request_sentry_logs_and_dump_to_file(url, token, mode, dates, out, chunk_size)
+    }
+}
+
+fn filter_logs_by_date(
+    logs: impl Iterator<Item = RawLog>,
+    dates: DateFilter,
+) -> impl Iterator<Item = RawLog> {
+    let DateFilter { from, to } = dates;
+
+    logs.take_while(move |l| {
```

Comment:
Does this work though? I think `take_while` stops after the first `false` value, so if the first log is before `from_date` no logs will be returned even though there are valid ones in the specified window

Reply:
Uhm, what do you mean? logs come sorted by date. `from_date` is the earliest one of them.

Reply:
Ah sorry, I thought they were sorted with increasing dates, but they are reversed